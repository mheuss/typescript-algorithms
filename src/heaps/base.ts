import { SelectComparisonFunction } from "../comparisons";
import { ComparisonFunction, Comparisons } from "../comparisons/constants";
import { IBaseHeapParameters } from "./types";

/**
 * Base heap constructor, has the functionality that is the basis of all heap types.
 */
export class BaseHeap<T> {
  protected data: T[];
  protected maxNumberOfChildren: number;
  protected comparisonFunction: ComparisonFunction<T> | undefined;
  protected minOrMax: Comparisons;

  constructor(options: IBaseHeapParameters<T>) {
    this.data = [];
    this.maxNumberOfChildren = options.maxNumberOfChildren;
    this.comparisonFunction = options.comparisonFunction;
    this.minOrMax = options.minOrMax;
  }

  /**
   * @returns Internal Queue Array
   */
  public getHeapContents = (): T[] => {
    return this.data;
  };

  /**
   * Empties heap
   */
  public empty = () => {
    this.data = [];
  };

  /**
   * @returns The number of items in the heap
   */
  public getLength = () => {
    return this.data.length;
  };

  /**
   * Pushed new items onto the heap
   * @param newValue
   */
  public push = (newValue: T) => {
    this.checkComparisonFunction([newValue]);

    this.data.push(newValue);
    this.bubbleUp(this.data.length - 1);
  };

  public pushArray = (newValues: T[]) => {
    newValues.forEach((value: T) => {
      this.push(value);
    });
  };

  /**
   * Removes an item from the top of the heap
   * @returns Topmost heap item, or undefined if none
   */
  public pop = (): T | undefined => {
    if (this.data.length === 0) {
      return undefined;
    }

    if (this.data.length === 1) {
      const rv = this.data[0];
      this.data = [];
      return rv;
    }

    this.checkComparisonFunction(this.data);

    const returnValue: T = this.data[0];
    this.data[0] = this.data.pop()!;

    this.bubbleDown(0);
    return returnValue;
  };

  private checkComparisonFunction = (incomingData: T[]): void => {
    if (this.comparisonFunction === undefined) {
      this.comparisonFunction = SelectComparisonFunction(incomingData);
    }
  };

  private bubbleDown = (index: number) => {
    if (index < 0 || index >= this.data.length) {
      return;
    }

    let largestChildIndex = this.maxNumberOfChildren * index + 1;

    for (let counter = 2; counter <= this.maxNumberOfChildren; counter++) {
      const childIndex = this.maxNumberOfChildren * index + counter;
      if (
        this.data[childIndex] !== undefined &&
        this.comparisonFunction!(
          this.data[childIndex],
          this.data[largestChildIndex]
        ) === this.minOrMax
      ) {
        largestChildIndex = childIndex;
      }
    }

    if (this.data[largestChildIndex] === undefined) {
      return;
    }

    if (
      this.comparisonFunction!(
        this.data[largestChildIndex],
        this.data[index]
      ) === this.minOrMax
    ) {
      const buffer = this.data[largestChildIndex];
      this.data[largestChildIndex] = this.data[index];
      this.data[index] = buffer;
      this.bubbleDown(largestChildIndex);
      return;
    }

    return;
  };

  private bubbleUp = (index: number) => {
    if (index < 1) {
      return;
    }

    const parentIndex = Math.floor((index - 1) / this.maxNumberOfChildren);
    if (
      this.comparisonFunction!(this.data[index], this.data[parentIndex]) ===
      this.minOrMax
    ) {
      const newIndex = this.data[parentIndex];
      this.data[parentIndex] = this.data[index];
      this.data[index] = newIndex;
    }

    this.bubbleUp(parentIndex);
  };
}
