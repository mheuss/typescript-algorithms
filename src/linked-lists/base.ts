import { ErrorCodes } from "../constants";
import { ListNode } from "./single-linked-list/list-node";
import {
  FindTypes,
  IFindType,
  SearchFunction,
} from "./single-linked-list/types";

export class LinkedList<T> {
  protected length: number;
  protected head: ListNode<T> | null;
  protected tail: ListNode<T> | null;

  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  /**
   * @returns Current length of list
   */
  public getLength = (): number => this.length;

  /**
   * Returns a linked list as an array.
   * @returns
   */
  public toArray = (): T[] => {
    const output: T[] = [];

    let currentNode = this.head;
    if (currentNode === null) {
      return output;
    }

    do {
      output.push(currentNode.payload);
      currentNode = currentNode.next;
    } while (currentNode !== null);

    return output;
  };

  /**
   * Uses the passed in function to find the first instance that matches, and returns that index and value that was matched against.
   * @param {SearchFunction} findComparison Function that contains the logic of the saerch. It is expected to return true if the
   * value of the item in the list matches what is being searched for - otherwise returns false.
   * @returns {IFindType | undefined} Either undefined if no matches, or Object contains the value of the matching item and the index it occupies in the list
   */
  public findFirst = (
    findComparison: SearchFunction<T>
  ): IFindType<T> | undefined => {
    let current = this.head;
    let index = 0;

    if (!current) {
      return undefined;
    }

    do {
      if (findComparison(current.payload)) {
        return {
          index,
          value: current.payload,
        };
      }
      index++;
      current = current.next;
    } while (current);

    return undefined;
  };

  /**
   * Finds all instances of matches found in a given list.
   * @param {SearchFunction} findComparison Function that contains the logic of the saerch. It is expected to return true if the
   * value of the item in the list matches what is being searched for - otherwise returns false.
   * @returns {FindTypes} Array of IFindType value. If no match, the array will be ewmpty
   */
  public findAll = (findComparison: SearchFunction<T>): FindTypes<T> => {
    let current = this.head;
    let index = 0;
    const matches: FindTypes<T> = [];

    if (!current) {
      return [];
    }

    do {
      if (findComparison(current.payload)) {
        matches.push({
          index,
          value: current.payload,
        });
      }
      index++;
      current = current.next;
    } while (current);

    return matches;
  };

  /**
   * Clears a list out
   * @returns Current Instance
   */
  public empty = (): LinkedList<T> => {
    this.length = 0;
    this.head = null;
    this.tail = null;
    return this;
  };

  /**
   * Returns the value of an item in the list
   * @param index Zero based index of an item in the list
   * @returns Value in that item
   */
  public get = (index: number): undefined | T => {
    const node = this.getNode(index);
    return !node ? undefined : node.payload;
  };

  /**
   * Sets a value at a given in the list
   * @param {number} index Zero based index of a given item in the list
   * @param {any} value The value this item at a given index will be set to
   * @returns {SingleLinkedList} The current SingleLinkedList instance
   */
  public set = (index: number, value: T): LinkedList<T> => {
    const node = this.getNode(index);
    if (!node) {
      throw {
        code: ErrorCodes.FAILED_TO_SET,
        message: "Failed to set the value at the given node",
      };
    }
    node.payload = value;
    return this;
  };

  public getNode = (index: number): null | ListNode<T> => {
    return this.getNodeInternal(index, this.length, this.head);
  };

  protected getNodeInternal(
    index: number,
    length: number,
    head: ListNode<T> | null
  ) {
    if (index >= length || head === null || index < 0) {
      return null;
    }

    if (index === 0) {
      return head;
    }

    let skips = 0;
    let current: ListNode<T> | null = head;
    do {
      if (current === null) {
        // Should never happen
        throw {
          code: ErrorCodes.LIST_IS_BROKEN_INTERNALLY,
          message: "List appears to be broken",
        };
      }
      current = current.next;
      skips++;
    } while (skips < index);

    return current;
  }
}
