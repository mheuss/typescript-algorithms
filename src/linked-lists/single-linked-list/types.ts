export type SearchFunction<T> = (value: T) => boolean;

export interface IFindType<T> {
  index: number;
  value: T;
}

export type FindTypes<T> = Array<IFindType<T>>;
