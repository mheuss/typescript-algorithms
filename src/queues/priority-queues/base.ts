import { MaxBinaryHeap } from "../../heaps/max-binary-heap";
import { MinBinaryHeap } from "../../heaps/min-binary-heap";
import { PriorityQueueNode } from "./priority-queue-node";

export class BasePriorityQueue<T> {
  protected heap: MaxBinaryHeap<PriorityQueueNode<T>>;

  constructor(
    heap:
      | MaxBinaryHeap<PriorityQueueNode<T>>
      | MinBinaryHeap<PriorityQueueNode<T>>
  ) {
    this.heap = heap;
  }

  public getLength = () => {
    return this.heap.getLength();
  };

  public enqueue = (value: T, priority: number) => {
    const node = new PriorityQueueNode<T>(value, priority);
    this.heap.push(node);
  };

  public dequeue = (): T | undefined => {
    const node = this.heap.pop();
    return node === undefined ? undefined : node.value;
  };

  public empty = () => {
    return this.heap.empty();
  };

  public getQueueContents = (): Array<PriorityQueueNode<T>> => {
    return this.heap.getHeapContents();
  };

  public setQueueContents = (value: Array<PriorityQueueNode<T>>) => {
    this.heap.pushArray(value);
  };
}
