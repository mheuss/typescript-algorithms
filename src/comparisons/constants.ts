export enum Comparisons {
  LessThan = -1,
  Equal,
  GreaterThan,
}

export type ComparisonFunction<T> = (a: T, b: T) => Comparisons;
