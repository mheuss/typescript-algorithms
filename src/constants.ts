export enum ErrorCodes {
  COMPARISON_FUNCTION_REQUIRED = 1,
  LIST_IS_BROKEN_INTERNALLY,
  OPERATION_BEYOND_BOUNDS,
  FAILED_TO_SET,
  COMPARISON_ERROR,
  INVALID_TREE_DATA_FILE,
}
