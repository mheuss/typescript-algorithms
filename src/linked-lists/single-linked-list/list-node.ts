export class ListNode<T> {
  public next: ListNode<T> | null = null;
  public payload: T;

  constructor(payload: T, previous?: ListNode<T>) {
    this.payload = payload;
    if (previous) {
      previous.next = this;
    } else {
      this.next = null;
    }
  }
}
