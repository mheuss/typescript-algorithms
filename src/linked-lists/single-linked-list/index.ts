import { ErrorCodes } from '../../constants';
import { LinkedList } from '../base';
import { ListNode } from './list-node';

/**
 * Single linked list implementation.
 * Accepts a generic for the payload that we want to carry in the list.
 */
export class SingleLinkedList<T> extends LinkedList<T> {
  /**
   * Removes from the list an item at position index, and returns that item.
   * @param {number} index of the item to remove.
   * @returns Value of the item removed
   * @throws {ErrorCodes.OPERATION_BEYOND_BOUNDS} If we pass in a bad index
   */
  public remove = (index: number): T | undefined => {
    if (index < 0 || index >= this.length) {
      throw {
        code: ErrorCodes.OPERATION_BEYOND_BOUNDS,
        message: 'Index is out of bounds',
      };
    }

    const pre = this.getNode(index - 1);

    if (pre === null) {
      return this.shift();
    }

    const node = pre.next;

    if (node === null) {
      throw {
        code: ErrorCodes.OPERATION_BEYOND_BOUNDS,
        message: 'Index is out of bounds',
      };
    }

    const post = node.next;

    if (post === null) {
      return this.pop();
    }

    pre.next = post;
    this.length--;

    return node.payload;
  };

  /**
   * Inserts a new node with the value passed in at the index provided.
   * @param index
   * @param value
   * @returns {SingleLinkedList} instance
   * @throws {ErrorCodes.OPERATION_BEYOND_BOUNDS} When index points beyond the length
   */
  public insert = (index: number, value: T): SingleLinkedList<T> => {
    if (index === 0) {
      this.unshift(value);
      return this;
    }

    if (index === this.length) {
      this.push(value);
      return this;
    }

    const preObject = this.getNode(index - 1);

    if (!preObject) {
      throw {
        code: ErrorCodes.OPERATION_BEYOND_BOUNDS,
        message: 'Operation beyond bounds',
      };
    }

    const postObject = preObject.next;
    if (!postObject) {
      this.push(value);
      return this;
    }

    const newNode = new ListNode<T>(value);
    preObject.next = newNode;
    newNode.next = postObject;
    this.length++;

    return this;
  };

  /**
   * Adds item to the beginning of the list
   * @param newItem
   * @returns {SingleLinkedList} The current SingleLinkedList instance
   */
  public unshift = (newItem: T): SingleLinkedList<T> => {
    const node = new ListNode(newItem);

    if (!this.head) {
      return this.push(newItem);
    }

    node.next = this.head;
    this.head = node;
    this.length++;

    return this;
  };

  /**
   * @returns value at the beginning of the array, or undefined if there is no value to return
   */
  public shift = (): T | undefined => {
    if (this.length === 0 || !this.head) {
      return undefined;
    }

    const itemToBeShifted = this.head;

    if (this.length === 1) {
      this.length = 0;
      this.head = null;
      this.tail = null;
      return itemToBeShifted.payload;
    }

    this.head = itemToBeShifted.next;
    this.length--;

    return itemToBeShifted.payload;
  };

  /**
   * Adds a single item of T to the end of the list.
   * @param newItem Item of T to be added
   * @returns The current SingleLinkedList instance
   */
  public push = (newItem: T): SingleLinkedList<T> => {
    const node = new ListNode(newItem);
    if (this.head === null) {
      this.head = node;
    }

    if (this.tail === null) {
      this.tail = node;
    } else {
      this.tail.next = node;
      this.tail = node;
    }

    this.length++;

    return this;
  };

  /**
   * Adds an array of T items to the end of the list, in array order.
   * @param newItem T[] to be added
   * @returns The current SingleLinkedList instance
   */
  public pushArray = (newItem: T[]) => {
    newItem.forEach(item => {
      this.push(item);
    });

    return this;
  };

  /**
   * Removes the last item on the list, and returns it
   * @returns Last Item on list, or undefined if no such item exists
   */
  public pop = (): T | undefined => {
    if (this.length === 0 || !this.tail) {
      return undefined;
    }

    const itemToBePopped = this.tail;

    if (this.length === 1) {
      this.empty();
      return itemToBePopped.payload;
    }

    const newLastItem = this.getNode(this.length - 2);
    if (newLastItem === null) {
      return undefined;
    }

    newLastItem.next = null;
    this.tail = newLastItem;
    this.length--;

    return itemToBePopped.payload;
  };

  /**
   * Reverses the list
   * @returns {SingleLinkedList} Instance
   */
  public reverse = (): SingleLinkedList<T> => {
    if (this.length <= 1) {
      return this;
    }
    const arr = this.toArray().reverse();
    this.empty();

    this.pushArray(arr);
    return this;
  };
}
