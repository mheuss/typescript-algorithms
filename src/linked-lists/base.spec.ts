import { ErrorCodes } from '../constants';
import { LinkedList } from './base';
import { ListNode } from './single-linked-list/list-node';

const HEAD = 'head';
const TAIL = 'tail';
const LENGTH = 'length';

describe('Linked List unit tests', () => {
  describe('Get Length', () => {
    it('Should return the list length', () => {
      const list = new LinkedList<number>();
      list[LENGTH] = 2;
      expect(list.getLength()).toEqual(2);
    });
  });

  describe('Empty', () => {
    it('Should empty a list', () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3]);
      list.empty();
      expect(list.getLength()).toEqual(0);
      expect(list[HEAD]).toBeNull();
      expect(list[TAIL]).toBeNull();
    });
  });

  describe('toArray', () => {
    it('Should return an empty array if given an empty list', () => {
      const list = new LinkedList<number>();
      expect(list.toArray()).toEqual([]);
    });

    it('Should return a list as an array', () => {
      const list = new LinkedList<number>();

      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7, 8, 9]);
      expect(list.toArray()).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9]);
    });
  });

  describe('findFirst', () => {
    it('Should return undefined with empty list', () => {
      const list = new LinkedList<number>();
      expect(list.findFirst(value => value === 7)).toBeUndefined();
    });

    it("Should return undefined if it can't find nuffin", () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7]);
      expect(list.findFirst(value => value === 32)).toBeUndefined();
    });

    it('Should return value if it matches - string', () => {
      const list = new LinkedList<string>();
      list[HEAD] = constructNodeList(['apple', 'pear', 'orange', 'banana']);
      expect(list.findFirst(value => value === 'pear')).toMatchObject({
        index: 1,
        value: 'pear',
      });
    });

    it('should return value if it matches, object', () => {
      const list = new LinkedList<any>();
      list[HEAD] = constructNodeList([
        { firstName: 'Mike' },
        { firstName: 'Bob' },
        { firstName: 'Jim' },
        { firstName: 'Ahtasham' },
        { firstName: 'Nigel' },
      ]);
      expect(list.findFirst(value => value.firstName === 'Jim')).toMatchObject({
        index: 2,
        value: { firstName: 'Jim' },
      });
    });

    it('Should return value if it matches, number, end of list', () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7]);
      expect(list.findFirst(value => value === 7)).toMatchObject({
        index: 6,
        value: 7,
      });
    });

    it('Should return value if it matches, number, beginning of list', () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7]);
      expect(list.findFirst(value => value < 2)).toMatchObject({
        index: 0,
        value: 1,
      });
    });
  });

  describe('findAll', () => {
    it('Should return undefined with empty list', () => {
      const list = new LinkedList<number>();
      expect(list.findAll(value => value === 7)).toEqual([]);
    });

    it("Should return undefined if it can't find nuffin", () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7]);
      expect(list.findAll(value => value === 32)).toEqual([]);
    });

    it('Should return value if it matches - string', () => {
      const list = new LinkedList<string>();
      list[HEAD] = constructNodeList(['apple', 'pear', 'orange', 'banana']);
      expect(list.findAll(value => value === 'pear')).toMatchObject([
        { index: 1, value: 'pear' },
      ]);
    });

    it('should return value if it matches, object', () => {
      const list = new LinkedList<any>();
      list[HEAD] = constructNodeList([
        { firstName: 'Mike' },
        { firstName: 'Bob' },
        { firstName: 'Jim' },
        { firstName: 'Ahtasham' },
        { firstName: 'Nigel' },
      ]);
      expect(list.findAll(value => value.firstName.length === 3)).toMatchObject(
        [
          { index: 1, value: { firstName: 'Bob' } },
          { index: 2, value: { firstName: 'Jim' } },
        ]
      );
    });

    it('Should return value if it matches, number, end of list', () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7]);
      expect(list.findAll(value => value === 7 || value === 3)).toMatchObject([
        { index: 2, value: 3 },
        { index: 6, value: 7 },
      ]);
    });

    it('Should return value if it matches, number, beginning of list', () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7]);
      expect(list.findAll(value => value < 2)).toMatchObject([
        { index: 0, value: 1 },
      ]);
    });
  });

  describe('Get', () => {
    it('Should return undefined if we are beyond the upper bounds ', () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7, 8, 9]);
      list[LENGTH] = 9;
      expect(list.get(300)).toBeUndefined();
    });

    it('Should return the value without entering the loop if index be all zero ', () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7, 8, 9]);
      list[LENGTH] = 9;
      expect(list.get(0)).toEqual(1);
    });

    it('Should throw if the list is internally broken ', () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7, 8, 9]);
      list[LENGTH] = 9;
      // @ts-ignore
      list[HEAD]?.next?.next = null;

      try {
        list.get(8);
        fail('Should have thrown');
      } catch (e) {
        expect(e.code).toEqual(ErrorCodes.LIST_IS_BROKEN_INTERNALLY);
      }
    });

    it('Should return the proper value ', () => {
      const list = new LinkedList<number>();
      list[HEAD] = constructNodeList([1, 2, 3, 4, 5, 6, 7, 8, 9]);
      list[LENGTH] = 9;
      expect(list.get(8)).toEqual(9);
      expect(list.get(2)).toEqual(3);
      expect(list.get(5)).toEqual(6);
    });
  });

  describe('Set', () => {
    it('Should return false when told to set beyond bounds', () => {
      const list = new LinkedList<number>();
      const arr: number[] = [1, 2, 3, 4, 5, 6, 7, 8, 9];
      list[HEAD] = constructNodeList(arr);
      list[LENGTH] = 9;
      try {
        list.set(100, 7);
        fail('Should have thrown');
      } catch (e) {
        expect(e.code).toEqual(ErrorCodes.FAILED_TO_SET);
      }
    });

    it('should return false when told to set on an empty list', () => {
      const list = new LinkedList<number>();
      try {
        list.set(100, 7);
        fail('Should have thrown');
      } catch (e) {
        expect(e.code).toEqual(ErrorCodes.FAILED_TO_SET);
      }
    });

    it('Should successfully set at a given point', () => {
      const list = new LinkedList<number>();
      const arr: number[] = [1, 2, 3, 4, 5, 6, 7, 8, 9];
      list[HEAD] = constructNodeList(arr);
      list[LENGTH] = 9;
      expect(list.set(2, 10)).toBeTruthy();
      expect(list.toArray()).toMatchSnapshot();
    });
  });
});

/*  ------------ Helper Functions ---------------*/

function constructNodeList<T>(incomingArray: T[]): ListNode<T> | null {
  if (incomingArray.length === 0) {
    return null;
  }

  const node = new ListNode<T>(incomingArray.shift() as T);
  addAdditionalNodes<T>(incomingArray, node);
  return node;
}

function addAdditionalNodes<T>(
  incomingArray: T[],
  previousNode: ListNode<T>
): ListNode<T> {
  if (incomingArray.length === 0) {
    return previousNode;
  }

  const node = new ListNode<T>(incomingArray.shift() as T, previousNode);
  return addAdditionalNodes<T>(incomingArray, node);
}
