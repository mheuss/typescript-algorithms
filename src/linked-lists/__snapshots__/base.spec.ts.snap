// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Linked List unit tests Set Should successfully set at a given point 1`] = `
Array [
  1,
  2,
  10,
  4,
  5,
  6,
  7,
  8,
  9,
]
`;
